The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3456285843/jobs/5768908133) [(plain)](https://github.com/rust-lang/rust/commit/89d4550737c194d1fd781d4db9c3c6d45b5f35b6/checks/9460572847/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling toml v0.5.9
error[E0308]: mismatched types
   --> doc.rs:570:12
    |
570 |         if builder.config.dry_run {
    |            ^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found enum `DryRun`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` due to previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:01:40
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
   --> doc.rs:570:12
    |
570 |         if builder.config.dry_run {
    |            ^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found enum `DryRun`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` due to previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:00:01
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
   --> doc.rs:570:12
    |
570 |         if builder.config.dry_run {
    |            ^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found enum `DryRun`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` due to previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:00:01
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
   --> doc.rs:570:12
    |
570 |         if builder.config.dry_run {
    |            ^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found enum `DryRun`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` due to previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:00:01
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
   --> doc.rs:570:12
    |
570 |         if builder.config.dry_run {
    |            ^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found enum `DryRun`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` due to previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:00:01
```

</details>

--- a/src/bootstrap/builder.rs
+++ b/src/bootstrap/builder.rs
@@ -1345,7 +1345,13 @@ impl<'a> Builder<'a> {
             let my_out = match mode {
                 // This is the intended out directory for compiler documentation.
                 Mode::Rustc | Mode::ToolRustc => self.compiler_doc_out(target),
-                Mode::Std => out_dir.join(target.triple).join("doc"),
+                Mode::Std => {
+                    if self.config.cmd.json() {
+                        out_dir.join(target.triple).join("json-doc")
+                    } else {
+                        out_dir.join(target.triple).join("doc")
+                    }
+                }
                 _ => panic!("doc mode {:?} not expected", mode),
             };
             let rustdoc = self.rustdoc(compiler);

--- a/src/bootstrap/doc.rs
+++ b/src/bootstrap/doc.rs
@@ -551,6 +551,49 @@ fn doc_std(
     extra_args: &[&OsStr],
     requested_crates: &[String],
 ) {
+    // `cargo` uses the same directory for both JSON docs and HTML docs.
+    // This could lead to cross-contamination when copying files into the specified `out` directory.
+    // For example:
+    // ```bash
+    // x doc std
+    // x doc std --json
+    // ```
+    // could lead to HTML docs being copied into the JSON docs output directory.
+    // To avoid this issue, we copy generated docs instead of whole directory by
+    // checking doc format and generated files.
+    fn cp_docs_by_doc_format(
+        format: &DocumentationFormat,
+        builder: &Builder<'_>,
+        src: &Path,
+        dst: &Path,
+    ) {
+        for f in builder.read_dir(src) {
+            let path = f.path();
+            let name = path.file_name().unwrap();
+            let dst = dst.join(name);
+
+            if t!(f.file_type()).is_dir() && format == &DocumentationFormat::HTML {
+                t!(fs::create_dir_all(&dst));
+                cp_docs_by_doc_format(format, builder, &path, &dst);
+            } else {
+                let _ = fs::remove_file(&dst);
+                let extension = path.extension().and_then(OsStr::to_str);
+
+                match format {
+                    DocumentationFormat::HTML if extension != Some("json") => {
+                        builder.copy(&path, &dst)
+                    }
+                    DocumentationFormat::JSON
+                        if extension == Some("json") || name.to_str() == Some(".stamp") =>
+                    {
+                        builder.copy(&path, &dst)
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+
     builder.info(&format!(
         "Documenting stage{} std ({}) in {} format",
         stage,
@@ -568,18 +611,6 @@ fn doc_std(
     // We will then copy the files from this directory into the final `out` directory, the specified
     // as a function parameter.
     let out_dir = builder.stage_out(compiler, Mode::Std).join(target.triple).join("doc");
-    // `cargo` uses the same directory for both JSON docs and HTML docs.
-    // This could lead to cross-contamination when copying files into the specified `out` directory.
-    // For example:
-    // ```bash
-    // x doc std
-    // x doc std --json
-    // ```
-    // could lead to HTML docs being copied into the JSON docs output directory.
-    // To avoid this issue, we clean the doc folder before invoking `cargo`.
-    if out_dir.exists() {
-        builder.remove_dir(&out_dir);
-    }
 
     let run_cargo_rustdoc_for = |package: &str| {
         let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "rustdoc");
@@ -605,7 +636,9 @@ fn doc_std(
         }
     }
 
-    builder.cp_r(&out_dir, &out);
+    if !builder.config.dry_run() {
+        cp_docs_by_doc_format(&format, builder, &out_dir, &out);
+    }
 }
 
 #[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]

--- a/src/bootstrap/doc.rs
+++ b/src/bootstrap/doc.rs
@@ -551,49 +551,6 @@ fn doc_std(
     extra_args: &[&OsStr],
     requested_crates: &[String],
 ) {
-    // `cargo` uses the same directory for both JSON docs and HTML docs.
-    // This could lead to cross-contamination when copying files into the specified `out` directory.
-    // For example:
-    // ```bash
-    // x doc std
-    // x doc std --json
-    // ```
-    // could lead to HTML docs being copied into the JSON docs output directory.
-    // To avoid this issue, we copy generated docs instead of whole directory by
-    // checking doc format and generated files.
-    fn cp_docs_by_doc_format(
-        format: &DocumentationFormat,
-        builder: &Builder<'_>,
-        src: &Path,
-        dst: &Path,
-    ) {
-        for f in builder.read_dir(src) {
-            let path = f.path();
-            let name = path.file_name().unwrap();
-            let dst = dst.join(name);
-
-            if t!(f.file_type()).is_dir() && format == &DocumentationFormat::HTML {
-                t!(fs::create_dir_all(&dst));
-                cp_docs_by_doc_format(format, builder, &path, &dst);
-            } else {
-                let _ = fs::remove_file(&dst);
-                let extension = path.extension().and_then(OsStr::to_str);
-
-                match format {
-                    DocumentationFormat::HTML if extension != Some("json") => {
-                        builder.copy(&path, &dst)
-                    }
-                    DocumentationFormat::JSON
-                        if extension == Some("json") || name.to_str() == Some(".stamp") =>
-                    {
-                        builder.copy(&path, &dst)
-                    }
-                    _ => {}
-                }
-            }
-        }
-    }
-
     builder.info(&format!(
         "Documenting stage{} std ({}) in {} format",
         stage,
@@ -607,15 +564,22 @@ fn doc_std(
         );
     }
     let compiler = builder.compiler(stage, builder.config.build);
+
+    let target_doc_dir_name = if format == DocumentationFormat::JSON { "json-doc" } else { "doc" };
+    let target_dir =
+        builder.stage_out(compiler, Mode::Std).join(target.triple).join(target_doc_dir_name);
+
     // This is directory where the compiler will place the output of the command.
     // We will then copy the files from this directory into the final `out` directory, the specified
     // as a function parameter.
-    let out_dir = builder.stage_out(compiler, Mode::Std).join(target.triple).join("doc");
+    let out_dir = target_dir.join(target.triple).join("doc");
 
     let run_cargo_rustdoc_for = |package: &str| {
         let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "rustdoc");
         compile::std_cargo(builder, target, compiler.stage, &mut cargo);
         cargo
+            .arg("--target-dir")
+            .arg(&*target_dir.to_string_lossy())
             .arg("-p")
             .arg(package)
             .arg("-Zskip-rustdoc-fingerprint")
@@ -636,9 +600,7 @@ fn doc_std(
         }
     }
 
-    if !builder.config.dry_run() {
-        cp_docs_by_doc_format(&format, builder, &out_dir, &out);
-    }
+    builder.cp_r(&out_dir, &out);
 }
 
 #[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]

--- a/src/test/run-make/rustdoc-verify-output-files/Makefile
+++ b/src/test/run-make/rustdoc-verify-output-files/Makefile
@@ -1,0 +1,36 @@
+include ../../run-make-fulldeps/tools.mk
+
+OUTPUT_DIR := "$(TMPDIR)/rustdoc"
+TMP_OUTPUT_DIR := "$(TMPDIR)/tmp-rustdoc"
+
+all:
+	# Generate html docs
+	$(RUSTDOC) src/lib.rs --crate-name foobar --crate-type lib --out-dir $(OUTPUT_DIR)
+
+	# Copy first output for to check if it's exactly same after second compilation
+	cp -R $(OUTPUT_DIR) $(TMP_OUTPUT_DIR)
+
+	# Generate html docs once again on same output
+	$(RUSTDOC) src/lib.rs --crate-name foobar --crate-type lib --out-dir $(OUTPUT_DIR)
+
+	# Check if everything exactly same
+	$(DIFF) -r -q $(OUTPUT_DIR) $(TMP_OUTPUT_DIR)
+
+	# Generate json doc on the same output
+	$(RUSTDOC) src/lib.rs --crate-name foobar --crate-type lib --out-dir $(OUTPUT_DIR) -Z unstable-options --output-format json
+
+	# Check if expected json file is generated
+	[ -e $(OUTPUT_DIR)/foobar.json ]
+
+	# TODO
+	# We should re-generate json doc once again and compare the diff with previously
+	# generated one. Because layout of json docs changes in each compilation, we can't
+	# do that currently.
+	#
+	# See https://github.com/rust-lang/rust/issues/103785#issuecomment-1307425590 for details.
+
+	# remove generated json doc
+	rm $(OUTPUT_DIR)/foobar.json
+
+	# Check if json doc compilation broke any of the html files generated previously
+	$(DIFF) -r -q $(OUTPUT_DIR) $(TMP_OUTPUT_DIR)

--- a/src/test/run-make/rustdoc-verify-output-files/src/lib.rs
+++ b/src/test/run-make/rustdoc-verify-output-files/src/lib.rs
@@ -1,0 +1,1 @@
+// nothing to see here

